core/services/run_queue.go

```diff
@@ -34,81 +35,176 @@ type RunQueue interface {
+	workers      map[models.ID]*singleJobSpecWorker
+	numWorkers   int32
+	chRuns       chan *models.JobRun
+	chStop       chan struct{}
+	chDone       chan struct{}
+	chWorkerDone chan models.ID
+}
 
-	runExecutor RunExecutor
+type singleJobSpecWorker struct {
+	chAddRun chan struct{}
+	chStop   chan struct{}
+	chDone   chan struct{}
 }
 
 // NewRunQueue initializes a RunQueue.
 func NewRunQueue(runExecutor RunExecutor) RunQueue {
 	return &runQueue{
-		workers:     make(map[string]int),
-		runExecutor: runExecutor,
+		workers:      make(map[models.ID]*singleJobSpecWorker),
+		runExecutor:  runExecutor,
+		chRuns:       make(chan *models.JobRun),
+		chStop:       make(chan struct{}),
+		chDone:       make(chan struct{}),
+		chWorkerDone: make(chan models.ID),
 	}
 }
 
 // Start prepares the job runner for accepting runs to execute.
 func (rq *runQueue) Start() error {
+	go rq.orchestrateWorkers()
 	return nil
 }
 
 // Stop closes all open worker channels.
 func (rq *runQueue) Stop() {
-	rq.workersMutex.Lock()
-	rq.stopRequested = true
-	rq.workersMutex.Unlock()
-	rq.workersWg.Wait()
+	close(rq.chStop)
+	<-rq.chDone
 }
 
 // Run tells the job runner to start executing a job
 func (rq *runQueue) Run(run *models.JobRun) {
-	rq.workersMutex.Lock()
-	if rq.stopRequested {
-		rq.workersMutex.Unlock()
-		return
+	select {
+	case rq.chRuns <- run:
+	case <-rq.chStop:
 	}
+}
+
+func (rq *runQueue) orchestrateWorkers() {
+	defer close(rq.chDone)
+	for {
+		select {
+		case run := <-rq.chRuns:
+			worker, exists := rq.workers[*run.ID]
+			if !exists {
+				worker = &singleJobSpecWorker{make(chan struct{}), make(chan struct{}), make(chan struct{})}
+				rq.workers[*run.ID] = worker
+				atomic.AddInt32(&rq.numWorkers, 1)
+				go rq.runSingleJobSpecWorker(run.ID, worker)
+				worker.chAddRun <- struct{}{}
+			} else {
+				select {
+				case worker.chAddRun <- struct{}{}:
+				case <-worker.chDone:
+					// If the worker is just spinning down as new work is coming in, allow it
+					// to die and spin up a new one.
+					worker = &singleJobSpecWorker{make(chan struct{}), make(chan struct{}), make(chan struct{})}
+					rq.workers[*run.ID] = worker
+					atomic.AddInt32(&rq.numWorkers, 1)
+					go rq.runSingleJobSpecWorker(run.ID, worker)
+					worker.chAddRun <- struct{}{}
+				}
+			}
+
+		case jobID := <-rq.chWorkerDone:
+			delete(rq.workers, jobID)
+			atomic.AddInt32(&rq.numWorkers, -1)
 
-	runID := run.ID.String()
-	defer numberRunsQueued.Inc()
-	if queueCount, present := rq.workers[runID]; present {
-		rq.workers[runID] = queueCount + 1
-		rq.workersMutex.Unlock()
-		return
+		case <-rq.chStop:
+			for _, run := range rq.workers {
+				run.chStop <- struct{}{}
+				<-run.chDone
+				atomic.AddInt32(&rq.numWorkers, -1)
+			}
+			return
+		}
 	}
-	rq.workers[runID] = 1
-	numberRunQueueWorkers.Set(float64(len(rq.workers)))
-	rq.workersMutex.Unlock()
+}
+
+func (rq *runQueue) runSingleJobSpecWorker(runID *models.ID, worker *singleJobSpecWorker) {
```

Comment:
I agree with [Code Climate](https://codeclimate.com/github/smartcontractkit/chainlink/pull/2399) that the [cognitive complexity](https://docs.codeclimate.com/docs/cognitive-complexity#section-the-idea) of this code is high. Breaking out the `worker` and coordinator loops into their own functions would help a little. Comments describing what the messages on each channel mean would help a lot.

Reply:
Agreed about cognitive complexity.  My thinking here, however, was that preserving locality of the small communicating worker processes inside of a single region of the code might actually be a little easier on the brain.  Then again, I wrote this (and am therefore not a reader), so I'll defer to others' opinions.  Can we wait for more feedback before choosing a direction?

Reply:
Sure.